### PR TITLE
add node template support

### PR DESCRIPTION
**Pull request by kon-angelo**
**How to categorize this PR?**

/area control-plane
/kind enhancement
/platform gcp

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
See https://github.com/gardener/gardener/issues/4948 for similar PRs

**Release note**:

```other operator
Provider-gcp now supports node template configuration for the workerpools. 
```

### CODE REVIEW

File: pkg/apis/gcp/validation/worker.go [hunk 1]
```diff
@@ -137,3 +141,32 @@ func validateDataVolume(workerConfig *gcp.WorkerConfig, volume core.DataVolume,
...
+		return allErrs
+	}
+
+	for _, capacityAttribute := range []corev1.ResourceName{"cpu", "gpu", "memory"} {
+		value, ok := nt.Capacity[capacityAttribute]
+		if !ok {
+			allErrs = append(allErrs, field.Required(fldPath.Child("nodeTemplate").Child("capacity"), fmt.Sprintf("%s is a mandatory field", capacityAttribute)))
+			continue
+		}
+		allErrs = append(allErrs, validateResourceQuantityValue(capacityAttribute, value, fldPath.Child("nodeTemplate").Child("capacity").Child(string(capacityAttribute)))...)
```
Comment (hebelsan) [inline pkg/apis/gcp/validation/worker.go:158]: same here.

File: pkg/controller/worker/machines.go [hunk 2]
```diff
@@ -246,15 +246,19 @@ func (w *workerDelegate) generateMachineConfig(_ context.Context) error {
...
-					Capacity:     initializeCapacity(pool.NodeTemplate.Capacity, gpuCount),
+					// always overwrite the GPU count if it was provided in the WorkerConfig.
+					Capacity:     initializeCapacity(nodeTemplate.Capacity, gpuCount),
 					InstanceType: pool.MachineType,
 					Region:       w.worker.Spec.Region,
 					Zone:         zone,
 				}
-
-				numGpus := pool.NodeTemplate.Capacity[ResourceGPU]
+				numGpus := workerConfig.NodeTemplate.Capacity[ResourceGPU]
```
Comment (hebelsan) [inline pkg/controller/worker/machines.go:261]: couldn't `workerConfig.NodeTemplate` be nil here?

**Review by hebelsan — CHANGES_REQUESTED** [review]

File: pkg/apis/gcp/validation/worker.go [hunk 3]
```diff
@@ -137,3 +141,32 @@ func validateDataVolume(workerConfig *gcp.WorkerConfig, volume core.DataVolume,
...
+	allErrs := field.ErrorList{}
+
+	if nt == nil {
+		return allErrs
+	}
+
+	for _, capacityAttribute := range []corev1.ResourceName{"cpu", "gpu", "memory"} {
+		value, ok := nt.Capacity[capacityAttribute]
+		if !ok {
+			allErrs = append(allErrs, field.Required(fldPath.Child("nodeTemplate").Child("capacity"), fmt.Sprintf("%s is a mandatory field", capacityAttribute)))
```
[the remark this answers is not in the document]
Comment (kon-angelo) [reply pkg/apis/gcp/validation/worker.go:155]: Good catch

**Review by hebelsan — APPROVED** [review]
/lgtm

**Review by hebelsan — APPROVED** [review]

**Pull request merged**